### PR TITLE
Add TSInstallAndQuit (parallel installation with post quit)

### PR DESCRIPTION
**Pull request by theHamsta**
Faster CI?

### CODE REVIEW

File: .github/workflows/check-query-files-and-compilation.yml [hunk 1]
```diff
@@ -47,12 +47,12 @@ jobs:
...
         if: matrix.os != 'windows-2022'
         run: |
-          nvim --headless -c "TSInstallSync all" -c "q"
+          nvim --headless -c "TSInstallAndQuit all" -c "sleep 180" -c "q"
 
       - name: Compile parsers Windows
         if: matrix.os == 'windows-2022'
         run: |
-          Neovim\\bin\\nvim.exe --headless -c "TSInstallSync all" -c "q"
+          Neovim\\bin\\nvim.exe  --headless "lua require'nvim-treesitter.install'.prefer_git=false" -c "TSInstallAndQuit all" -c "sleep 180" -c "q"
```
Comment (theHamsta) [inline .github/workflows/check-query-files-and-compilation.yml:55]: also curl is faster than git clone of a whole repo

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: why doesn't Windows use curl even now?

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: was a typo: `-c` was missing

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: but installing tsx and typescript in parallel confuses Windows :cry:

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: and Windows uses too much heap space on CI :smile:

File: .github/workflows/check-query-files-and-compilation.yml [hunk 2]
```diff
@@ -47,12 +47,12 @@ jobs:
       - name: Compile parsers Unix like
         if: matrix.os != 'windows-2022'
         run: |
-          nvim --headless -c "TSInstallSync all" -c "q"
+          nvim --headless -c "TSInstallAndQuit all" -c "sleep 180" -c "q"
```
Comment (kyazdani42) [inline .github/workflows/check-query-files-and-compilation.yml:50]: isn't the CI going to take too long with this sleep ?

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml:50]: No. Sleep is not executed fully, when quit happens earlier.

File: .github/workflows/check-query-files-and-compilation.yml (hunk 1 above)
Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: Our problem is that this also seems to happen right now with curl and ocaml

File: .github/workflows/check-query-files-and-compilation.yml (hunk 2 above)
Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml:50]:
```
❯ time nvim --headless -c "TSInstallAndQuit jsdoc comment" -c "sleep 180" -c "q"
[nvim-treesitter] [0/2] Downloading...
[nvim-treesitter] [0/2] Downloading...
[nvim-treesitter] [0/2] Creating temporary directory
[nvim-treesitter] [0/2] Extracting...
[nvim-treesitter] [0/2] Compiling...
[nvim-treesitter] [0/2] Creating temporary directory
[nvim-treesitter] [0/2] Extracting...
[nvim-treesitter] [0/2] Compiling...
[nvim-treesitter] [1/2] Treesitter parser for comment has been installednvim --headless -c "TSInstallAndQuit jsdoc comment" -c "sleep 180" -c "q"
 0,94s user 0,23s system 96% cpu 1,214 total
```

**Review by kyazdani42 — APPROVED** [review]

**Review by vigoux — APPROVED** [review]

**Review by stsewd — APPROVED** [review]

**Comment by jzelinskie** [comment]
Wouldn't a better solution be copying how [packer does this](https://github.com/wbthomason/packer.nvim/blob/db3c3e3379613443d94e677a6ac3f61278e36e47/lua/packer.lua#L327-L330) by creating a "TSInstallComplete" autocommand so that the net result would be ```nvim --headless -c 'autocmd User TSInstallComplete quitall' -c 'TSInstall'```?

**Comment by vigoux** [comment]
I kind of like your idea better @jzelinskie

**Comment by clason** [comment]
If you `git clone`, you should also add `--depth=1` to speed up the clone (especially on older repos with a long history).

**Comment by theHamsta** [comment]
@clason git is not used. We just curl the commit we need. Except on Windows. This would also change that.

You can only `--depth=1` branches while we're using raw commits. Also it is not guaranteed that git checkout will find that commit in case the commit in no longer in the history of any upstream branch.

**Comment by clason** [comment]
superseded by rewrite on `main` branch (which has full async)

**Pull request closed without merging**